### PR TITLE
evmbump2

### PR DESCRIPTION
**Pull request by xwvvvvwx**
this takes @MrChico's pull request (https://github.com/dapphub/klab/pull/283) and bumps the `evm-semantics` to the latest release ([v1.0.0-3efa855](https://github.com/kframework/evm-semantics/releases/tag/v1.0.0-3efa855)) with a few of our changes on top (branch: [dapphub/evm-semantics@klab-new](https://github.com/dapphub/evm-semantics/tree/klab-new)).

I also swapped the submodule target from [MrChico/evm-semnatics](https://github.com/MrChico/evm-semantics/) to [dapphub/evm-semantics](https://github.com/dapphub/evm-semantics/) as I don't have write access to the @MrChico fork :).

Opening a draft PR now for some advice: In general what kind of testing should be performed here to shake out any potential issues? I can already see a few failures when I run `make test` locally, but I was thinking something like more extensive a passing run of the `k-dss` specs would be appropriate? Or does that not make sense?

Finally I was wondering about the status of the llvm and haskell backends, are they ready to go? Should we be thinking of switching away from the Java backend?

### CODE REVIEW

**Comment by ehildenb** [comment]
If you're using the OCaml backend (which I guess not, since KLab only uses Java backend), then you should switch that usage to LLVM backend.

Haskell backend does not support the StateLog functionality that KLab takes advantage of. It supports its own repl which allows exploring the execution graph, but it's really more targeted at people who are developing the Haskell backend.

Haskell backend can currently pass some simple ERC20 proofs in time comparable to Java backend, so we're in the process of migrating all the EVM proofs over. My guess is another 2 months.

Haskell backend is also potentially going to add structured logging, kinda like StateLog. Instead of logging everything to disk though, like StateLog, it will be injected into an SQL database which can be queried. The same database can be used for resuming proofs and such too, ideally we'll make a "statelog wrapper" around that to make it work with KLab too.

In short, stay with Java backend.

**Comment by xwvvvvwx** [comment]
thx for the detailed answer @ehildenb! :pray: :sparkles:

File: .gitmodules [hunk 1]
```diff
@@ -1,4 +1,4 @@
 [submodule "evm-semantics"]
 	path = evm-semantics
-	url = https://github.com/mrchico/evm-semantics
+	url = https://github.com/dapphub/evm-semantics
```
Comment (asymmetric) [inline .gitmodules]: Why not point it to upstream?

Comment (xwvvvvwx) [reply .gitmodules]: we have a couple of small patches applied on top.

**Comment by asymmetric** [comment]
Note: I had to run `git submodule sync -- evm-semantics` after checking out this branch, as it changes the submodule url.

**Comment by xwvvvvwx** [comment]
> Note: I had to run git submodule sync -- evm-semantics after checking out this branch, as it changes the submodule url.

Is that maybe why the CI is failing @mhhf ?

**Comment by xwvvvvwx** [comment]
so `klab prove` on `Vat_heal_pass_rough` in `examples/heal` seems to run into a infinite loop of `keccak` -> `keccakIntList` ([full stderr](https://cloudflare-ipfs.com/ipfs/Qmf2B9q8Um2N3CdEhpxjwLGXzWvzdFKwCivdYCE6dqwQiB)). 

I see [this change](https://github.com/kframework/evm-semantics/pull/589/files#diff-1ef0db86ab0e9570588bbd121f32a26dL329) and I'm wondering if it's related? I'm not really sure how to debug further, any tips @MrChico @livnev @mhhf ?

**Comment by ehildenb** [comment]
@xwvvvvwx it probably is related to the change you linked. The new Java backend allows passing a `--concrete-rules` file, so that for each proof you can specify which rules are treated as having `concrete` attribute.

So you'll need a file which has one rule name per line (in this case `keccakIntList`), and pass `--concrete-rules NAME_OF_FILE` to it. Then it will treat that rule as before.

See KEVM build system for more examples, the proofs there have been converted over to `--concrete-rules` approach.

**Comment by xwvvvvwx** [comment]
@ehildenb thanks so much! I added the `concrete-rules` file but it didn't seem to help. I still get the same infinite loop as before :/ 

I wasn't really sure what should be in the file, so I just copied the one from [here](https://github.com/kframework/evm-semantics/blob/master/tests/specs/concrete-rules.txt). Was that the right thing to do?

**Comment by ehildenb** [comment]
It looks like your file contains EDSL.keccakIntList, so that's good. But I wonder if you need to mark more/less as concrete.

**Comment by ehildenb** [comment]
I'm sorry, I made a mistake. You need to pass a comma-separate list of the rule names, not a filename containing the rules, to the `--concrete-rules` option.

**Comment by asymmetric** [comment]
Diff of the same spec, before and after the bump:

```diff
4c4
< module 678DB80C5D2CB7B3BCCD4547E53DA72021EEF51C7631BC4557A32FCFDEC55A80
---
> module B75A3DB66E15A4FAA1F5D21078D63C34C1CFCA0F949B9AEFAAF4894526AA34CD
24,25c24,25
<         <program> #asMapOpCodes(#dasmOpCodes(Vat_bin_runtime, PETERSBURG)) </program>
<         <programBytes> Vat_bin_runtime </programBytes>
---
>         <program> Vat_bin_runtime </program>
>         <jumpDests> #computeValidJumpDests(Vat_bin_runtime) </jumpDests>
46,62c46,64
<       <previousHash> _ </previousHash>
<       <ommersHash> _ </ommersHash>
<       <coinbase> _ </coinbase>
<       <stateRoot> _ </stateRoot>
<       <transactionsRoot> _ </transactionsRoot>
<       <receiptsRoot> _ </receiptsRoot>
<       <logsBloom> _ </logsBloom>
<       <difficulty> _ </difficulty>
<       <number> BLOCK_NUMBER </number>
<       <gasLimit> _ </gasLimit>
<       <gasUsed> _ </gasUsed>
<       <timestamp> TIME </timestamp>
<       <extraData> _ </extraData>
<       <mixHash> _ </mixHash>
<       <blockNonce> _ </blockNonce>
<       <ommerBlockHeaders> _ </ommerBlockHeaders>
<       <blockhash> _ </blockhash>
---
>       <blockhashes> _ </blockhashes>
>       <block>
>         <previousHash> _ </previousHash>
>         <ommersHash> _ </ommersHash>
>         <coinbase> _ </coinbase>
>         <stateRoot> _ </stateRoot>
>         <transactionsRoot> _ </transactionsRoot>
>         <receiptsRoot> _ </receiptsRoot>
>         <logsBloom> _ </logsBloom>
>         <difficulty> _ </difficulty>
>         <number> _ </number>
>         <gasLimit> _ </gasLimit>
>         <gasUsed> _ </gasUsed>
>         <timestamp> TIME </timestamp>
>         <extraData> _ </extraData>
>         <mixHash> _ </mixHash>
>         <blockNonce> _ </blockNonce>
>         <ommerBlockHeaders> _ </ommerBlockHeaders>
>       </block>
178c180
<   andBool (#sizeWordStack(CD) <=Int 1250000000
---
>   andBool (#sizeByteArray(CD) <=Int 1250000000
```

File: libexec/klab-prove [hunk 2]
```diff
@@ -108,13 +108,6 @@ concrete_rules=(
   EVM.#lookup.some
   EVM.#lookup.none
   EVM.#memoryUsageUpdate.some
```
Comment (asymmetric) [inline libexec/klab-prove]: I wonder if the # get parsed as comments :thinking:

**Comment by xwvvvvwx** [comment]
assuming that this should be closed in favor of: https://github.com/dapphub/klab/pull/318 ?

File: libexec/klab-prove (hunk 2 above)
Comment (asymmetric) [reply libexec/klab-prove]: Nope.

Comment (asymmetric) [reply libexec/klab-prove]: Nope
